apps/full-site-editing/full-site-editing-plugin/full-site-editing/templates/class-template-image-inserter.php

```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * Full site editing file.
+ *
+ * @package A8C\FSE
+ */
+
+namespace A8C\FSE;
+
+/**
+ * Class Template_Image_Inserter
+ */
+class Template_Image_Inserter {
+	/**
+	 * Uploads specific images to the media libraray
+	 *
+	 * It then checks if any of the passed image URLs are also in the post content
+	 * of any of the passed post IDs. If so, then the post content is updated to
+	 * use the new local URL instead of the passed URL.
+	 *
+	 * @param array $img_urls URLs of images which should be uploaded.
+	 * @param array $post_ids An array of posts IDs which should be updated with newly uploaded image URLs.
+	 */
+	public function copy_images_and_update_posts( $img_urls = [], $post_ids = [] ) {
+		do_action(
+			'a8c_fse_log',
+			'template_image_sideload_begin',
+			[
+				'context'    => 'Template_Image_Inserter->copy_images_and_update_posts',
+				'image_urls' => $img_urls,
+			]
+		);
+
+		if ( empty( $img_urls ) || empty( $post_ids ) ) {
+			return;
+		}
+
+		$images = $this->upload_images( $img_urls );
+		foreach ( $post_ids as $id ) {
+			$this->update_post_images( $id, $images );
+		}
+	}
+
+	/**
+	 * Upload an array of images to the media library.
+	 *
+	 * If an file fails to upload, the error is logged to the FSE error handler.
+	 *
+	 * @param  array $image_urls Images URLs to upload.
+	 * @return array Each entry contains the old and new URL of an image if successfully uploaded.
+	 */
+	public function upload_images( $image_urls ) {
+		return array_reduce(
+			$image_urls,
+			function( $accumulator, $url ) {
+				// 1. Attempt to sideload the image and get the local URL.
+				$local_url = $this->upload_image( $url );
+
+				// 2. Check if anything went wrong during upload.
+				$error_msg = null;
+				if ( is_wp_error( $local_url ) ) {
+					$error_msg = $local_url->get_error_message();
+				} elseif ( ! is_string( $local_url ) ) {
+					$error_msg = 'Issue uploading the image.';
+				}
+
+				// 3. If there was an error, report it.
+				if ( is_string( $error_msg ) ) {
+					do_action(
+						'a8c_fse_log',
+						'image_sideload_failure',
+						[
+							'context'          => 'Template_Image_Inserter->upload_images',
+							'error'            => $error_msg,
+							'remote_image_url' => $url,
+						]
+					);
+				} else {
+					// 4. Otherwise, save the url.
+					$accumulator[] = [
+						'old_url' => $url,
+						'new_url' => $local_url,
+					];
+				}
+				return $accumulator;
+			},
+			[]
+		);
+	}
+
+	/**
+	 * Sideload a remote image URL and return the local attachment URL.
+	 *
+	 * @param  string $image_url The URL to sideload to the local site.
+	 * @return string The local URL of the newly uploaded image.
+	 */
+	public function upload_image( $image_url ) {
+		if ( ! function_exists( 'media_handle_sideload' ) ) {
+			require_once ABSPATH . '/wp-admin/includes/file.php';
+			require_once ABSPATH . '/wp-admin/includes/media.php';
+			require_once ABSPATH . '/wp-admin/includes/image.php';
+		}
```

Comment:
Any reasons why including all three of these?
AFAICS `media_handle_sideload` is in `media.php`.
https://core.trac.wordpress.org/browser/tags/5.2/src/wp-admin/includes/media.php#L419

Reply:
Oh, ok, I see that `download_url` is defined in `file.php`, but what about `image.php` then?

Reply:
To be honest, I copy-pasted this from the headstart class which does this. (though I did change a lot of stuff to make it clearer and just use what we need.) I imagine these imports aren't really needed... but does it hurt?